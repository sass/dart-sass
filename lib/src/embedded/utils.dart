// Copyright 2019 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'dart:typed_data';

import 'package:protobuf/protobuf.dart';
import 'package:source_span/source_span.dart';
import 'package:term_glyph/term_glyph.dart' as term_glyph;

import '../syntax.dart';
import 'embedded_sass.pb.dart' as proto;
import 'embedded_sass.pb.dart' hide SourceSpan, Syntax;
import 'util/varint_builder.dart';

/// The special ID that indicates an error that's not associated with a specific
/// inbound request ID.
const errorId = 0xffffffff;

/// Returns a [ProtocolError] indicating that a mandatory field with the given
/// [fieldName] was missing.
ProtocolError mandatoryError(String fieldName) =>
    paramsError("Missing mandatory field $fieldName");

/// Returns a [ProtocolError] indicating that the parameters for an inbound
/// message were invalid.
ProtocolError paramsError(String message) => ProtocolError()
  // Set the ID to [errorId] by default. This will be overwritten by the
  // dispatcher if a request ID is available.
  ..id = errorId
  ..type = ProtocolErrorType.PARAMS
  ..message = message;

/// Returns a [ProtocolError] with type `PARSE` and the given [message].
ProtocolError parseError(String message) => ProtocolError()
  ..type = ProtocolErrorType.PARSE
  ..message = message;

/// Converts a Dart source span to a protocol buffer source span.
proto.SourceSpan protofySpan(SourceSpan span) {
  var protoSpan = proto.SourceSpan()
    ..text = span.text
    ..start = _protofyLocation(span.start)
    ..end = _protofyLocation(span.end)
    ..url = span.sourceUrl?.toString() ?? "";
  if (span is SourceSpanWithContext) protoSpan.context = span.context;
  return protoSpan;
}

/// Converts a Dart source location to a protocol buffer source location.
SourceSpan_SourceLocation _protofyLocation(SourceLocation location) =>
    SourceSpan_SourceLocation()
      ..offset = location.offset
      ..line = location.line
      ..column = location.column;

/// Converts a protocol buffer syntax enum into a Sass API syntax enum.
Syntax syntaxToSyntax(proto.Syntax syntax) {
  switch (syntax) {
    case proto.Syntax.SCSS:
      return Syntax.scss;
    case proto.Syntax.INDENTED:
      return Syntax.sass;
    case proto.Syntax.CSS:
      return Syntax.css;
    default:
      throw "Unknown syntax $syntax.";
  }
}

/// Returns the result of running [callback] with the global ASCII config set
/// to [ascii].
T withGlyphs<T>(T callback(), {required bool ascii}) {
  var currentConfig = term_glyph.ascii;
  term_glyph.ascii = ascii;
  var result = callback();
  term_glyph.ascii = currentConfig;
  return result;
}

/// Serializes [value] to an unsigned varint.
Uint8List serializeVarint(int value) {
  if (value == 0) return Uint8List.fromList([0]);
  RangeError.checkNotNegative(value);

  // Essentially `(value.bitLength / 7).ceil()`, but without getting floats
  // involved.
  var lengthInBytes = (value.bitLength + 6) ~/ 7;
  var list = Uint8List(lengthInBytes);
  for (var i = 0; i < lengthInBytes; i++) {
    // The highest-order bit indicates whether more bytes are necessary to fully
    // express the number. The lower 7 bits indicate the number's value.
    list[i] = (value > 0x7f ? 0x80 : 0) | (value & 0x7f);
    value >>= 7;
  }
  return list;
}

/// Serializes a compilation ID and protobuf message into a packet buffer as
/// specified in the embedded protocol.
Uint8List serializePacket(int compilationId, GeneratedMessage message) {
  var varint = serializeVarint(compilationId);
  var protobufWriter = CodedBufferWriter();
  message.writeToCodedBufferWriter(protobufWriter);

  var packet = Uint8List(varint.length + protobufWriter.lengthInBytes);
  packet.setAll(0, varint);
  protobufWriter.writeTo(packet, varint.length);
  return packet;
}

/// A [VarintBuilder] that's shared across invocations of [parsePacket] to avoid
/// unnecessary allocations.
final _compilationIdBuilder = VarintBuilder(32, 'compilation ID');

/// Parses a compilation ID and encoded protobuf message from a packet buffer as
/// specified in the embedded protocol.
(int, Uint8List) parsePacket(Uint8List packet) {
  try {
    var i = 0;
    while (true) {
      if (i == packet.length) {
        throw parseError(
            "Invalid compilation ID: continuation bit always set.");
      }

      var compilationId = _compilationIdBuilder.add(packet[i]);
      i++;
      if (compilationId != null) {
        return (compilationId, Uint8List.sublistView(packet, i));
      }
    }
  } finally {
    _compilationIdBuilder.reset();
  }
}
