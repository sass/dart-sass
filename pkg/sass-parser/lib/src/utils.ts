// Copyright 2024 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import * as postcss from 'postcss';

import {Node} from './node';

/**
 * A type that matches any constructor for {@link T}. From
 * https://www.typescriptlang.org/docs/handbook/mixins.html.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Constructor<T> = new (...args: any[]) => T;

/**
 * An explicit field description passed to `cloneNode` that describes in detail
 * how to clone it.
 */
interface ExplicitClonableField<Name extends string> {
  /** The field's name. */
  name: Name;

  /**
   * Whether the field can be set to an explicit undefined value which means
   * something different than an absent field.
   */
  explicitUndefined?: boolean;
}

/** The type of field names that can be passed into `cloneNode`. */
type ClonableField<Name extends string> = Name | ExplicitClonableField<Name>;

/** Makes a {@link ClonableField} explicit. */
function parseClonableField<Name extends string>(
  field: ClonableField<Name>
): ExplicitClonableField<Name> {
  return typeof field === 'string' ? {name: field} : field;
}

/**
 * Creates a copy of {@link node} by passing all the properties in {@link
 * constructorFields} as an object to its constructor.
 *
 * If {@link overrides} is passed, it overrides any existing constructor field
 * values. It's also used to assign {@link assignedFields} after the cloned
 * object has been constructed.
 */
export function cloneNode<T extends Pick<postcss.Node, 'source'>>(
  node: T,
  overrides: Record<string, unknown> | undefined,
  constructorFields: ClonableField<keyof T & string>[],
  assignedFields?: ClonableField<keyof T & string>[]
): T {
  // We have to do these casts because the actual `...Prop` types that get
  // passed in and used for the constructor aren't actually subtypes of
  // `Partial<T>`. They use `never` types to ensure that various properties are
  // mutually exclusive, which is not compatible.
  const typedOverrides = overrides as Partial<T> | undefined;
  const constructorFn = node.constructor as new (defaults: Partial<T>) => T;

  const constructorParams: Partial<T> = {};
  for (const field of constructorFields) {
    const {name, explicitUndefined} = parseClonableField(field);
    let value: T[keyof T & string] | undefined;
    if (
      typedOverrides &&
      (explicitUndefined
        ? Object.hasOwn(typedOverrides, name)
        : typedOverrides[name] !== undefined)
    ) {
      value = typedOverrides[name];
    } else {
      value = maybeClone(node[name]);
    }
    if (value !== undefined) constructorParams[name] = value;
  }
  const cloned = new constructorFn(constructorParams);

  if (typedOverrides && assignedFields) {
    for (const field of assignedFields) {
      const {name, explicitUndefined} = parseClonableField(field);
      if (
        explicitUndefined
          ? Object.hasOwn(typedOverrides, name)
          : typedOverrides[name]
      ) {
        // This isn't actually guaranteed to be non-null, but TypeScript
        // (correctly) complains that we could be passing an undefined value to
        // a field that doesn't allow undefined. We don't have a good way of
        // forbidding that while still allowing users to override values that do
        // explicitly allow undefined, though.
        cloned[name] = typedOverrides[name]!;
      }
    }
  }

  cloned.source = node.source;
  return cloned;
}

/**
 * If {@link value} is a Sass node, a record, or an array, clones it and returns
 * the clone. Otherwise, returns it as-is.
 */
function maybeClone<T>(value: T): T {
  if (Array.isArray(value)) return value.map(maybeClone) as T;
  if (typeof value !== 'object' || value === null) return value;
  // The only records we care about are raws, which only contain primitives and
  // arrays of primitives, so structued cloning is safe.
  if (value.constructor === Object) return structuredClone(value);
  if (value instanceof postcss.Node) return value.clone() as T;
  return value;
}

/**
 * Converts {@link node} into a JSON-safe object, with the given {@link fields}
 * included.
 *
 * This always includes the `type`, `sassType`, `raws`, and `source` fields if
 * set. It converts multiple references to the same source input object into
 * indexes into a top-level list.
 */
export function toJSON<T extends Node>(
  node: T,
  fields: (keyof T & string)[],
  inputs?: Map<postcss.Input, number>
): object {
  // Only include the inputs field at the top level.
  const includeInputs = !inputs;
  inputs ??= new Map();
  let inputIndex = inputs.size;

  const result: Record<string, unknown> = {};
  if ('type' in node) result.type = (node as {type: string}).type;

  fields = ['sassType', 'raws', ...fields];
  for (const field of fields) {
    const value = node[field];
    if (value !== undefined) result[field] = toJsonField(field, value, inputs);
  }

  if (node.source) {
    let inputId = inputs.get(node.source.input);
    if (inputId === undefined) {
      inputId = inputIndex++;
      inputs.set(node.source.input, inputId);
    }

    result.source = {
      start: node.source.start,
      end: node.source.end,
      inputId,
    };
  }

  if (includeInputs) {
    result.inputs = [...inputs.keys()].map(input => input.toJSON());
  }
  return result;
}

/**
 * Converts a single field with name {@link field} and value {@link value} to a
 * JSON-safe object.
 *
 * The {@link inputs} map works the same as it does in {@link toJSON}.
 */
function toJsonField(
  field: string,
  value: unknown,
  inputs: Map<postcss.Input, number>
): unknown {
  if (typeof value !== 'object' || value === null) {
    return value;
  } else if (Array.isArray(value)) {
    return value.map((element, i) =>
      toJsonField(i.toString(), element, inputs)
    );
  } else if ('toJSON' in value) {
    if ('sassType' in value) {
      return (
        value as {
          toJSON: (field: string, inputs: Map<postcss.Input, number>) => object;
        }
      ).toJSON('', inputs);
    } else {
      return (value as {toJSON: (field: string) => object}).toJSON(field);
    }
  } else {
    return value;
  }
}

/**
 * Returns the longest string (of code units) that's an initial substring of
 * every string in
 * {@link strings}.
 */
export function longestCommonInitialSubstring(strings: string[]): string {
  let candidate: string | undefined;
  for (const string of strings) {
    if (candidate === undefined) {
      candidate = string;
    } else {
      for (let i = 0; i < candidate.length && i < string.length; i++) {
        if (candidate.charCodeAt(i) !== string.charCodeAt(i)) {
          candidate = candidate.substring(0, i);
          break;
        }
      }
      candidate = candidate.substring(
        0,
        Math.min(candidate.length, string.length)
      );
    }
  }
  return candidate ?? '';
}
