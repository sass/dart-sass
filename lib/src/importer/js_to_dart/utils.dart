// Copyright 2023 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:node_interop/js.dart';

import '../../js/utils.dart';
import '../utils.dart';

/// Throws a JsError if [scheme] isn't a valid URL scheme.
void validateUrlScheme(String scheme) {
  if (!isValidUrlScheme(scheme)) {
    jsThrow(
        JsError('"$scheme" isn\'t a valid URL scheme (for example "file").'));
  }
}
