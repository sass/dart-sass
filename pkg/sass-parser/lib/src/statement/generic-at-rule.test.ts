// Copyright 2024 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import {GenericAtRule, Interpolation, Root, Rule, css, sass, scss} from '../..';
import * as utils from '../../../test/utils';

describe('a generic @-rule', () => {
  let node: GenericAtRule;
  describe('with no children', () => {
    describe('with no params', () => {
      function describeNode(
        description: string,
        create: () => GenericAtRule
      ): void {
        describe(description, () => {
          beforeEach(() => void (node = create()));

          it('has type atrule', () => expect(node.type).toBe('atrule'));

          it('has sassType atrule', () => expect(node.sassType).toBe('atrule'));

          it('has a nameInterpolation', () =>
            expect(node).toHaveInterpolation('nameInterpolation', 'foo'));

          it('has a name', () => expect(node.name).toBe('foo'));

          it('has no paramsInterpolation', () =>
            expect(node.paramsInterpolation).toBeUndefined());

          it('has empty params', () => expect(node.params).toBe(''));

          it('has undefined nodes', () => expect(node.nodes).toBeUndefined());
        });
      }

      describeNode(
        'parsed as SCSS',
        () => scss.parse('@foo').nodes[0] as GenericAtRule
      );

      describeNode(
        'parsed as CSS',
        () => css.parse('@foo').nodes[0] as GenericAtRule
      );

      describeNode(
        'parsed as Sass',
        () => sass.parse('@foo').nodes[0] as GenericAtRule
      );

      describe('constructed manually', () => {
        describeNode(
          'with a name interpolation',
          () =>
            new GenericAtRule({
              nameInterpolation: new Interpolation({nodes: ['foo']}),
            })
        );

        describeNode(
          'with a name string',
          () => new GenericAtRule({name: 'foo'})
        );
      });

      describe('constructed from ChildProps', () => {
        describeNode('with a name interpolation', () =>
          utils.fromChildProps({
            nameInterpolation: new Interpolation({nodes: ['foo']}),
          })
        );

        describeNode('with a name string', () =>
          utils.fromChildProps({name: 'foo'})
        );
      });
    });

    describe('with params', () => {
      function describeNode(
        description: string,
        create: () => GenericAtRule
      ): void {
        describe(description, () => {
          beforeEach(() => void (node = create()));

          it('has a name', () => expect(node.name.toString()).toBe('foo'));

          it('has a paramsInterpolation', () =>
            expect(node).toHaveInterpolation('paramsInterpolation', 'bar'));

          it('has matching params', () => expect(node.params).toBe('bar'));

          it('has undefined nodes', () => expect(node.nodes).toBeUndefined());
        });
      }

      describeNode(
        'parsed as SCSS',
        () => scss.parse('@foo bar').nodes[0] as GenericAtRule
      );

      describeNode(
        'parsed as CSS',
        () => css.parse('@foo bar').nodes[0] as GenericAtRule
      );

      describeNode(
        'parsed as Sass',
        () => sass.parse('@foo bar').nodes[0] as GenericAtRule
      );

      describe('constructed manually', () => {
        describeNode(
          'with an interpolation',
          () =>
            new GenericAtRule({
              name: 'foo',
              paramsInterpolation: new Interpolation({nodes: ['bar']}),
            })
        );

        describeNode(
          'with a param string',
          () => new GenericAtRule({name: 'foo', params: 'bar'})
        );
      });

      describe('constructed from ChildProps', () => {
        describeNode('with an interpolation', () =>
          utils.fromChildProps({
            name: 'foo',
            paramsInterpolation: new Interpolation({nodes: ['bar']}),
          })
        );

        describeNode('with a param string', () =>
          utils.fromChildProps({name: 'foo', params: 'bar'})
        );
      });
    });
  });

  describe('with empty children', () => {
    describe('with no params', () => {
      function describeNode(
        description: string,
        create: () => GenericAtRule
      ): void {
        describe(description, () => {
          beforeEach(() => void (node = create()));

          it('has a name', () => expect(node.name).toBe('foo'));

          it('has no paramsInterpolation', () =>
            expect(node.paramsInterpolation).toBeUndefined());

          it('has empty params', () => expect(node.params).toBe(''));

          it('has no nodes', () => expect(node.nodes).toHaveLength(0));
        });
      }

      describeNode(
        'parsed as SCSS',
        () => scss.parse('@foo {}').nodes[0] as GenericAtRule
      );

      describeNode(
        'parsed as CSS',
        () => css.parse('@foo {}').nodes[0] as GenericAtRule
      );

      describeNode(
        'constructed manually',
        () => new GenericAtRule({name: 'foo', nodes: []})
      );

      describeNode('constructed from ChildProps', () =>
        utils.fromChildProps({name: 'foo', nodes: []})
      );
    });

    describe('with params', () => {
      function describeNode(
        description: string,
        create: () => GenericAtRule
      ): void {
        describe(description, () => {
          beforeEach(() => void (node = create()));

          it('has a name', () => expect(node.name.toString()).toBe('foo'));

          it('has a paramsInterpolation', () =>
            expect(node).toHaveInterpolation('paramsInterpolation', 'bar '));

          it('has matching params', () => expect(node.params).toBe('bar '));
        });
      }

      describeNode(
        'parsed as SCSS',
        () => scss.parse('@foo bar {}').nodes[0] as GenericAtRule
      );

      describeNode(
        'parsed as CSS',
        () => css.parse('@foo bar {}').nodes[0] as GenericAtRule
      );

      describe('constructed manually', () => {
        describeNode(
          'with params',
          () =>
            new GenericAtRule({
              name: 'foo',
              params: 'bar ',
              nodes: [],
            })
        );

        describeNode(
          'with an interpolation',
          () =>
            new GenericAtRule({
              name: 'foo',
              paramsInterpolation: new Interpolation({nodes: ['bar ']}),
              nodes: [],
            })
        );
      });

      describe('constructed from ChildProps', () => {
        describeNode('with params', () =>
          utils.fromChildProps({
            name: 'foo',
            params: 'bar ',
            nodes: [],
          })
        );

        describeNode('with an interpolation', () =>
          utils.fromChildProps({
            name: 'foo',
            paramsInterpolation: new Interpolation({nodes: ['bar ']}),
            nodes: [],
          })
        );
      });
    });
  });

  describe('with a child', () => {
    describe('with no params', () => {
      describe('parsed as Sass', () => {
        beforeEach(() => {
          node = sass.parse('@foo\n  .bar').nodes[0] as GenericAtRule;
        });

        it('has a name', () => expect(node.name).toBe('foo'));

        it('has no paramsInterpolation', () =>
          expect(node.paramsInterpolation).toBeUndefined());

        it('has empty params', () => expect(node.params).toBe(''));

        it('has a child node', () => {
          expect(node.nodes).toHaveLength(1);
          expect(node.nodes[0]).toBeInstanceOf(Rule);
          expect(node.nodes[0]).toHaveProperty('selector', '.bar\n');
        });
      });
    });

    describe('with params', () => {
      function describeNode(
        description: string,
        create: () => GenericAtRule
      ): void {
        describe(description, () => {
          beforeEach(() => void (node = create()));

          it('has a name', () => expect(node.name.toString()).toBe('foo'));

          it('has a paramsInterpolation', () =>
            expect(node).toHaveInterpolation('paramsInterpolation', 'bar'));

          it('has matching params', () => expect(node.params).toBe('bar'));

          it('has a child node', () => {
            expect(node.nodes).toHaveLength(1);
            expect(node.nodes[0]).toBeInstanceOf(Rule);
            expect(node.nodes[0]).toHaveProperty('selector', '.baz\n');
          });
        });
      }

      describeNode(
        'parsed as Sass',
        () => sass.parse('@foo bar\n  .baz').nodes[0] as GenericAtRule
      );

      describe('constructed manually', () => {
        describeNode(
          'with params',
          () =>
            new GenericAtRule({
              name: 'foo',
              params: 'bar',
              nodes: [{selector: '.baz\n'}],
            })
        );
      });

      describe('constructed from ChildProps', () => {
        describeNode('with params', () =>
          utils.fromChildProps({
            name: 'foo',
            params: 'bar',
            nodes: [{selector: '.baz\n'}],
          })
        );
      });
    });
  });

  describe('assigned new name', () => {
    beforeEach(() => {
      node = scss.parse('@foo {}').nodes[0] as GenericAtRule;
    });

    it("removes the old name's parent", () => {
      const oldName = node.nameInterpolation!;
      node.nameInterpolation = 'bar';
      expect(oldName.parent).toBeUndefined();
    });

    it("assigns the new interpolation's parent", () => {
      const interpolation = new Interpolation({nodes: ['bar']});
      node.nameInterpolation = interpolation;
      expect(interpolation.parent).toBe(node);
    });

    it('assigns the interpolation explicitly', () => {
      const interpolation = new Interpolation({nodes: ['bar']});
      node.nameInterpolation = interpolation;
      expect(node.nameInterpolation).toBe(interpolation);
    });

    it('assigns the interpolation as a string', () => {
      node.nameInterpolation = 'bar';
      expect(node).toHaveInterpolation('nameInterpolation', 'bar');
    });

    it('assigns the interpolation as name', () => {
      node.name = 'bar';
      expect(node).toHaveInterpolation('nameInterpolation', 'bar');
    });
  });

  describe('assigned new params', () => {
    beforeEach(() => {
      node = scss.parse('@foo bar {}').nodes[0] as GenericAtRule;
    });

    it('removes the old interpolation', () => {
      node.paramsInterpolation = undefined;
      expect(node.paramsInterpolation).toBeUndefined();
    });

    it('removes the old interpolation as undefined params', () => {
      node.params = undefined;
      expect(node.params).toBe('');
      expect(node.paramsInterpolation).toBeUndefined();
    });

    it('removes the old interpolation as empty string params', () => {
      node.params = '';
      expect(node.params).toBe('');
      expect(node.paramsInterpolation).toBeUndefined();
    });

    it("removes the old interpolation's parent", () => {
      const oldParams = node.paramsInterpolation!;
      node.paramsInterpolation = undefined;
      expect(oldParams.parent).toBeUndefined();
    });

    it("assigns the new interpolation's parent", () => {
      const interpolation = new Interpolation({nodes: ['baz']});
      node.paramsInterpolation = interpolation;
      expect(interpolation.parent).toBe(node);
    });

    it('assigns the interpolation explicitly', () => {
      const interpolation = new Interpolation({nodes: ['baz']});
      node.paramsInterpolation = interpolation;
      expect(node.paramsInterpolation).toBe(interpolation);
    });

    it('assigns the interpolation as a string', () => {
      node.paramsInterpolation = 'baz';
      expect(node).toHaveInterpolation('paramsInterpolation', 'baz');
    });

    it('assigns the interpolation as params', () => {
      node.params = 'baz';
      expect(node).toHaveInterpolation('paramsInterpolation', 'baz');
    });
  });

  describe('stringifies', () => {
    describe('to SCSS', () => {
      describe('with undefined nodes', () => {
        describe('without params', () => {
          it('with default raws', () =>
            expect(new GenericAtRule({name: 'foo'}).toString()).toBe('@foo;'));

          it('with afterName', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {afterName: '/**/'},
              }).toString()
            ).toBe('@foo/**/;'));

          it('with afterName', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {afterName: '/**/'},
              }).toString()
            ).toBe('@foo/**/;'));

          it('with between', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {between: '/**/'},
              }).toString()
            ).toBe('@foo/**/;'));

          it('with afterName and between', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {afterName: '/*afterName*/', between: '/*between*/'},
              }).toString()
            ).toBe('@foo/*afterName*//*between*/;'));
        });

        describe('with params', () => {
          it('with default raws', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                paramsInterpolation: 'baz',
              }).toString()
            ).toBe('@foo baz;'));

          it('with afterName', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                paramsInterpolation: 'baz',
                raws: {afterName: '/**/'},
              }).toString()
            ).toBe('@foo/**/baz;'));

          it('with between', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                paramsInterpolation: 'baz',
                raws: {between: '/**/'},
              }).toString()
            ).toBe('@foo baz/**/;'));
        });

        it('with after', () =>
          expect(
            new GenericAtRule({name: 'foo', raws: {after: '/**/'}}).toString()
          ).toBe('@foo;'));

        it('with before', () =>
          expect(
            new Root({
              nodes: [new GenericAtRule({name: 'foo', raws: {before: '/**/'}})],
            }).toString()
          ).toBe('/**/@foo'));
      });

      describe('with defined nodes', () => {
        describe('without params', () => {
          it('with default raws', () =>
            expect(new GenericAtRule({name: 'foo', nodes: []}).toString()).toBe(
              '@foo {}'
            ));

          it('with afterName', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {afterName: '/**/'},
                nodes: [],
              }).toString()
            ).toBe('@foo/**/ {}'));

          it('with afterName', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {afterName: '/**/'},
                nodes: [],
              }).toString()
            ).toBe('@foo/**/ {}'));

          it('with between', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {between: '/**/'},
                nodes: [],
              }).toString()
            ).toBe('@foo/**/{}'));

          it('with afterName and between', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {afterName: '/*afterName*/', between: '/*between*/'},
                nodes: [],
              }).toString()
            ).toBe('@foo/*afterName*//*between*/{}'));
        });

        describe('with params', () => {
          it('with default raws', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                paramsInterpolation: 'baz',
                nodes: [],
              }).toString()
            ).toBe('@foo baz {}'));

          it('with afterName', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                paramsInterpolation: 'baz',
                raws: {afterName: '/**/'},
                nodes: [],
              }).toString()
            ).toBe('@foo/**/baz {}'));

          it('with between', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                paramsInterpolation: 'baz',
                raws: {between: '/**/'},
                nodes: [],
              }).toString()
            ).toBe('@foo baz/**/{}'));
        });

        describe('with after', () => {
          it('with no children', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                raws: {after: '/**/'},
                nodes: [],
              }).toString()
            ).toBe('@foo {/**/}'));

          it('with a child', () =>
            expect(
              new GenericAtRule({
                name: 'foo',
                nodes: [{selector: '.bar'}],
                raws: {after: '/**/'},
              }).toString()
            ).toBe('@foo {\n    .bar {}/**/}'));
        });

        it('with before', () =>
          expect(
            new Root({
              nodes: [
                new GenericAtRule({
                  name: 'foo',
                  raws: {before: '/**/'},
                  nodes: [],
                }),
              ],
            }).toString()
          ).toBe('/**/@foo {}'));
      });
    });
  });

  describe('clone', () => {
    let original: GenericAtRule;
    beforeEach(() => {
      original = scss.parse('@foo bar {.baz {}}').nodes[0] as GenericAtRule;
      // TODO: remove this once raws are properly parsed
      original.raws.between = '  ';
    });

    describe('with no overrides', () => {
      let clone: GenericAtRule;
      beforeEach(() => void (clone = original.clone()));

      describe('has the same properties:', () => {
        it('nameInterpolation', () =>
          expect(clone).toHaveInterpolation('nameInterpolation', 'foo'));

        it('name', () => expect(clone.name).toBe('foo'));

        it('params', () => expect(clone.params).toBe('bar '));

        it('paramsInterpolation', () =>
          expect(clone).toHaveInterpolation('paramsInterpolation', 'bar '));

        it('raws', () => expect(clone.raws).toEqual({between: '  '}));

        it('source', () => expect(clone.source).toBe(original.source));

        it('nodes', () => {
          expect(clone.nodes).toHaveLength(1);
          expect(clone.nodes[0]).toBeInstanceOf(Rule);
          expect(clone.nodes[0]).toHaveProperty('selector', '.baz ');
        });
      });

      describe('creates a new', () => {
        it('self', () => expect(clone).not.toBe(original));

        for (const attr of [
          'nameInterpolation',
          'paramsInterpolation',
          'raws',
          'nodes',
        ] as const) {
          it(attr, () => expect(clone[attr]).not.toBe(original[attr]));
        }
      });

      describe('sets parent for', () => {
        it('nodes', () => expect(clone.nodes[0].parent).toBe(clone));
      });
    });

    describe('overrides', () => {
      describe('name', () => {
        describe('defined', () => {
          let clone: GenericAtRule;
          beforeEach(() => {
            clone = original.clone({name: 'qux'});
          });

          it('changes name', () => expect(clone.name).toBe('qux'));

          it('changes nameInterpolation', () =>
            expect(clone).toHaveInterpolation('nameInterpolation', 'qux'));
        });

        describe('undefined', () => {
          let clone: GenericAtRule;
          beforeEach(() => {
            clone = original.clone({name: undefined});
          });

          it('preserves name', () => expect(clone.name).toBe('foo'));

          it('preserves nameInterpolation', () =>
            expect(clone).toHaveInterpolation('nameInterpolation', 'foo'));
        });
      });

      describe('nameInterpolation', () => {
        describe('defined', () => {
          let clone: GenericAtRule;
          beforeEach(() => {
            clone = original.clone({
              nameInterpolation: new Interpolation({nodes: ['qux']}),
            });
          });

          it('changes name', () => expect(clone.name).toBe('qux'));

          it('changes nameInterpolation', () =>
            expect(clone).toHaveInterpolation('nameInterpolation', 'qux'));
        });

        describe('undefined', () => {
          let clone: GenericAtRule;
          beforeEach(() => {
            clone = original.clone({nameInterpolation: undefined});
          });

          it('preserves name', () => expect(clone.name).toBe('foo'));

          it('preserves nameInterpolation', () =>
            expect(clone).toHaveInterpolation('nameInterpolation', 'foo'));
        });
      });

      describe('raws', () => {
        it('defined', () =>
          expect(original.clone({raws: {afterName: '  '}}).raws).toEqual({
            afterName: '  ',
          }));

        it('undefined', () =>
          expect(original.clone({raws: undefined}).raws).toEqual({
            between: '  ',
          }));
      });

      describe('params', () => {
        describe('defined', () => {
          let clone: GenericAtRule;
          beforeEach(() => {
            clone = original.clone({params: 'qux'});
          });

          it('changes params', () => expect(clone.params).toBe('qux'));

          it('changes paramsInterpolation', () =>
            expect(clone).toHaveInterpolation('paramsInterpolation', 'qux'));
        });

        describe('undefined', () => {
          let clone: GenericAtRule;
          beforeEach(() => {
            clone = original.clone({params: undefined});
          });

          it('changes params', () => expect(clone.params).toBe(''));

          it('changes paramsInterpolation', () =>
            expect(clone.paramsInterpolation).toBeUndefined());
        });
      });

      describe('paramsInterpolation', () => {
        describe('defined', () => {
          let clone: GenericAtRule;
          let interpolation: Interpolation;
          beforeEach(() => {
            interpolation = new Interpolation({nodes: ['qux']});
            clone = original.clone({paramsInterpolation: interpolation});
          });

          it('changes params', () => expect(clone.params).toBe('qux'));

          it('changes paramsInterpolation', () =>
            expect(clone).toHaveInterpolation('paramsInterpolation', 'qux'));
        });

        describe('undefined', () => {
          let clone: GenericAtRule;
          beforeEach(() => {
            clone = original.clone({paramsInterpolation: undefined});
          });

          it('changes params', () => expect(clone.params).toBe(''));

          it('changes paramsInterpolation', () =>
            expect(clone.paramsInterpolation).toBeUndefined());
        });
      });
    });
  });

  describe('toJSON', () => {
    it('without params', () =>
      expect(scss.parse('@foo').nodes[0]).toMatchSnapshot());

    it('with params', () =>
      expect(scss.parse('@foo bar').nodes[0]).toMatchSnapshot());

    it('with empty children', () =>
      expect(scss.parse('@foo {}').nodes[0]).toMatchSnapshot());

    it('with a child', () =>
      expect(scss.parse('@foo {@bar}').nodes[0]).toMatchSnapshot());
  });
});
