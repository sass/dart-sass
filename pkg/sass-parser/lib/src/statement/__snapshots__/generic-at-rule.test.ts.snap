// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a generic @-rule toJSON with a child 1`] = `
{
  "inputs": [
    {
      "css": "@foo {@bar}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "name": "foo",
  "nameInterpolation": <foo>,
  "nodes": [
    <@bar;>,
  ],
  "params": "",
  "raws": {},
  "sassType": "atrule",
  "source": <1:1-1:12 in 0>,
  "type": "atrule",
}
`;

exports[`a generic @-rule toJSON with empty children 1`] = `
{
  "inputs": [
    {
      "css": "@foo {}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "name": "foo",
  "nameInterpolation": <foo>,
  "nodes": [],
  "params": "",
  "raws": {},
  "sassType": "atrule",
  "source": <1:1-1:8 in 0>,
  "type": "atrule",
}
`;

exports[`a generic @-rule toJSON with params 1`] = `
{
  "inputs": [
    {
      "css": "@foo bar",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "name": "foo",
  "nameInterpolation": <foo>,
  "params": "bar",
  "paramsInterpolation": <bar>,
  "raws": {},
  "sassType": "atrule",
  "source": <1:1-1:9 in 0>,
  "type": "atrule",
}
`;

exports[`a generic @-rule toJSON without params 1`] = `
{
  "inputs": [
    {
      "css": "@foo",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "name": "foo",
  "nameInterpolation": <foo>,
  "params": "",
  "raws": {},
  "sassType": "atrule",
  "source": <1:1-1:5 in 0>,
  "type": "atrule",
}
`;
